### PR TITLE
GEOMESA-788 Add Proper Partitioning Control To GeoSpark

### PR DESCRIPTION
**Pull request by JakeKurzer**
This is a messy mid-term solution to the over-partitioning issue. It does not resolve the under-partitioning issue, which would require finagling the actual ranges. I didn't feel comfortable enough with the deadlines or the codebases to pursue that, but we do need to discuss it.

Coding style's a little rough, but I've tried to be mostly conformant. The added API function is deprecated, as it will be merged into the main API function down the road, once we solve the under-partitioning problem.

### CODE REVIEW

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 1]
```diff
@@ -22,25 +22,67 @@ import java.util.UUID
...
+      numShards = IndexSchema.maxShard(ds.getIndexSchemaFmt(sft.getTypeName))
+     }
+
+    override def getSplits(context: JobContext) : java.util.List[InputSplit] = {
+      init(context.getConfiguration)
+      val desiredSplits = context.getConfiguration.getInt("SplitsDesired", 10*numShards)
+      val accumuloSplits = delegate.getSplits(context)
+      // try to create 2 mappers per node - account for case where there are less splits than shards
+      val groupSize = Math.max(1, accumuloSplits.length / desiredSplits)
+      logger.error("I am being used....")
```
Comment (elahrvivaz) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: ...

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 2]
```diff
@@ -22,25 +22,67 @@ import java.util.UUID
...
+      val params = GeoMesaConfigurator.getDataStoreInParams(conf)
+      val ds = DataStoreFinder.getDataStore(params).asInstanceOf[AccumuloDataStore]
+      sft = ds.getSchema(GeoMesaConfigurator.getFeatureType(conf))
+      encoding = ds.getFeatureEncoding(sft)
+      numShards = IndexSchema.maxShard(ds.getIndexSchemaFmt(sft.getTypeName))
+     }
+
+    override def getSplits(context: JobContext) : java.util.List[InputSplit] = {
+      init(context.getConfiguration)
+      val desiredSplits = context.getConfiguration.getInt("SplitsDesired", 10*numShards)
```
Comment (elahrvivaz) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: maybe we should just push this 'desired splits' option into the GeoMesaInputFormat? Seems like the rest of the logic is duplicated from there?

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 3]
```diff
@@ -55,6 +97,15 @@ object GeoMesaSpark {
   def typeProp(typeName: String) = s"geomesa.types.$typeName"
   def jOpt(typeName: String, spec: String) = s"-D${typeProp(typeName)}=$spec"
 
+  @deprecated("Relies on incidental functional parity across classes, effectively an implicit and unexpressed dependency on non-guaranteed behavior", "Apr2015")
+  def rddPartPerExec(conf: Configuration, sc: SparkContext, dsParams: Map[String, String], query: Query, partsPerExec: Int): RDD[SimpleFeature] = {
```
Comment (elahrvivaz) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: [style] lots of long lines

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala (hunk 2 above)
Comment (JakeKurzer) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: At the moment, but I think it'll probably diverge a lot further over time.

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 4]
```diff
@@ -55,6 +97,15 @@ object GeoMesaSpark {
   def typeProp(typeName: String) = s"geomesa.types.$typeName"
   def jOpt(typeName: String, spec: String) = s"-D${typeProp(typeName)}=$spec"
 
+  @deprecated("Relies on incidental functional parity across classes, effectively an implicit and unexpressed dependency on non-guaranteed behavior", "Apr2015")
```
Comment (elahrvivaz) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: so you're adding a function that you've already deprecated? maybe this should be a TODO or something?

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala (hunk 1 above)
Comment (JakeKurzer) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: Oops, left my debug in. It wasn't clear whether mapred or mapreduce was being used, so I was worried that there might be name resolution issues.

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 5]
```diff
@@ -22,25 +22,67 @@ import java.util.UUID
...
+          val split = new GroupedSplit()
+          split.location = location
+          split.splits.append(group.map(_.asInstanceOf[RangeInputSplit]): _*)
+          split
+        }
+      }
+      //we've dumped all ranges into groups by location, and tried to get enough splits by breaking up those groups.
+      //we may eventually need to go ahead and do some pretty grim mojo to break each split down small enough
+      //so that we don't blow executor's memory pools. Right now, that's not happening, so I'm not going to push it.
+      logger.error(s"Got ${splits.toList.length} splits using desired=${desiredSplits} from ${accumuloSplits.length}")
```
Comment (elahrvivaz) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: should be debug probably

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala (hunk 4 above)
Comment (JakeKurzer) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: Probably should be TODO, but it's used exclusively externally, so I went with deprecated. I do plan to patch this thing out of existence in the next month or so.

**Comment by JakeKurzer** [comment]
@elahrvivaz I removed the.... uhh... extraneous logging.

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala (hunk 5 above)
Comment (JakeKurzer) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: went with info. see newest.

File: geomesa-compute/pom.xml [hunk 6]
```diff
@@ -41,6 +41,16 @@
             <scope>compile</scope>
         </dependency>
         <dependency>
+            <groupId>org.locationtech.geomesa</groupId>
+            <artifactId>geomesa-jobs-accumulo1.5</artifactId>
+            <exclusions>
```
Comment (jnh5y) [inline geomesa-compute/pom.xml]: What's the reason for the exclusion?

Comment (JakeKurzer) [inline geomesa-compute/pom.xml]: Kryo version conflict, if I remember correctly. It's been a bit.

Comment (JakeKurzer) [inline geomesa-compute/pom.xml]: It's the same pattern that @elahrvivaz used in a couple other places, I think. .... I think?

**Comment by JakeKurzer** [comment]
@elahrvivaz fixed the lines. check it? I'd like to get this merged in.

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 7]
```diff
@@ -55,25 +97,71 @@ object GeoMesaSpark {
...
+      ds.connector.whoami(),
+      ds.authToken)
 
-    if(useMock) ConfiguratorBase.setMockInstance(classOf[AccumuloInputFormat], conf, ds.connector.getInstance().getInstanceName)
-    else ConfiguratorBase.setZooKeeperInstance(classOf[AccumuloInputFormat], conf, ds.connector.getInstance().getInstanceName, ds.connector.getInstance().getZooKeepers)
+    if(useMock) {
+      ConfiguratorBase.setMockInstance(classOf[AccumuloInputFormat],
+        conf,
+        ds.connector.getInstance().getInstanceName)
+    }else
```
Comment (jnh5y) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: (Style) Space after the }.

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala [hunk 8]
```diff
@@ -55,25 +97,71 @@ object GeoMesaSpark {
...
+    {ConfiguratorBase.setZooKeeperInstance(classOf[AccumuloInputFormat],
```
Comment (jnh5y) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: (Style) 
```
} else { 
  ....
```

File: geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala (hunk 2 above)
Comment (jnh5y) [inline geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala]: I'm with Emilio. 

Let's try out the simple fix.

**Comment by elahrvivaz** [comment]
@jnh5y are you good with it now?

**Comment by elahrvivaz** [comment]
@JakeKurzer sorry, I've been out...

**Comment by JakeKurzer** [comment]
@elahrvivaz tomorrow I'm going to code up some handling for extremely large ranges, so that'll probably make the divergence more justified. Let's plan to talk prior to symposium tomorrow?

**Pull request closed without merging**